### PR TITLE
new(proposals): API versioning for user/kernel boundary

### PR DESCRIPTION
**Pull request by gnosek**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

/kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area driver-kmod

> /area driver-ebpf

> /area libscap

> /area libsinsp

> /area tests

/area proposals

**What this PR does / why we need it**:

This proposal introduces semver-compatible version checks for the user/kernel boundary,
i.e. between the kernel driver/eBPF probe and the userspace components.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: proposals/20210818-driver-semver.md
Comment (deepskyblue86) [inline proposals/20210818-driver-semver.md:5]:
```suggestion
This proposal introduces [semver](https://semver.org/)-compatible version checks for the user/kernel boundary,
```

**Review by deepskyblue86 — DISMISSED** [review]

**Review by leogr — COMMENTED** [review]
Excellent proposal. Just a few tips on naming and one thought regarding point 4.

Also, I guess we will start with version 1.0.0? If so, it would be preferable to specify it in the proposal.

Btw, curious to know users' feedback :eyes: 

Thank you!

File: proposals/20210818-driver-semver.md
Comment (araujof) [reply proposals/20210818-driver-semver.md:82]: This proposal addresses a lot of usability and maintenance issues for API consumers. I like the idea of a default/canonical name (`scap`). 

I lean towards a parametric build, where the consumer name can be passed to override a default consumer name.

Comment (gnosek) [reply proposals/20210818-driver-semver.md:82]: I'm fine with leaving an option to specify the module name as an internal I-know-what-I'm-doing hack but IMO it shouldn't be the default way of building. I'll try to update the wording to better express the idea.

**Comment by gnosek** [comment]
If we rename the driver, we can start at 1.0.0. Otherwise we need to start with a version higher than all known consumers (I proposed 32 in the original PR, did anybody go higher than that?).

We'll probably have to phase out PROBE_VERSION some time after this proposal gets implemented so that consumers can switch to the new scheme on their own schedule. Maybe just make it optional.

File: proposals/20210818-driver-semver.md
Comment (terylt) [reply proposals/20210818-driver-semver.md:82]: This is a nice proposal that is really needed to support libsinsp and libscap as separate entities from their consumers.   The probe should definitely come with a different consumer agnostic name, and that name and version should be used in the pre-built probes that are downloadable to make maintenance easier.  

As for whether the name should be parameterized, I don't have a strong opinion.  I think most consumers are going to want to take advantage of the pre-built probes they can download; however, there may be cases where two different consumers want to have two different probes.  As such, I think probably making the name configurable but defaulted, wouldn't add a lot of complexity to the build process, but give flexibility to change for those who really want too.  

Finally, one thing that I think might go really well with this proposal would be to:
1.) move a version of the falco-driver-loader and docker-entrypoint.sh script into the libraries, as these are  key scripts for any consumer using the probe.

**Comment by zuc** [comment]
@gnosek I do really like the proposed approach, however I'm thinking whether this should come along with a proper test suite (e.g. to reduce risk of unintentional BC introduction), WDYT?

File: proposals/20210818-driver-semver.md
Comment (gnosek) [reply proposals/20210818-driver-semver.md]: Moving `falco-driver-loader` into this repo would be a good idea, *but* it wouldn't really be a standalone script then (we don't want to hardcode the Falco driver repo for all consumers).

Maybe it would work to make two scripts: one to build the driver and another one to download a prebuilt driver (with the logic to determine the config hash etc.). Consumers could then wrap these two however they want and provide the prebuilt drivers from wherever they want.

Not sure about the docker-entrypoint script. There will have to be some per-consumer logic (at the very minimum, we're starting different processes in the end) so maybe it would work as a shell library to be sourced into the consumers' entry point scripts.

**Comment by gnosek** [comment]
@zuc:

Do I think having decent test coverage is a good idea? Yes.
Do I believe it's in scope for the proposal? Not really (we'd want tests regardless of this proposal).
Do I have the cycles to write tests for the whole user/kernel API? I wish.

**Comment by zuc** [comment]
@gnosek to better frame my thinking: if this goes through and we end up "centralizing" drivers, relevance of a proper test suite changes dramatically IMO

**Comment by terylt** [comment]
> If we rename the driver, we can start at 1.0.0. Otherwise we need to start with a version higher than all known consumers (I proposed 32 in the original PR, did anybody go higher than that?).
> 
> We'll probably have to phase out PROBE_VERSION some time after this proposal gets implemented so that consumers can switch to the new scheme on their own schedule. Maybe just make it optional.
 
We used the libs in our sysflow collector, but the collector pretended to be sysdig, and more recently falco, for versioning so we could take advantage of the precompiled probes.  I imagine most other consumers would do something similar to take advantage of that feature.

**Comment by gnosek** [comment]
@terylt I'd hope everybody switches to `scap` as the probe name and nobody will need to masquerade as anybody else :) As a bonus, you won't need to determine the sysdig/falco version your code is compatible with

@zuc sure, but does that affect this proposal?

**Comment by terylt** [comment]
> @terylt I'd hope everybody switches to `scap` as the probe name and nobody will need to masquerade as anybody else :) As a bonus, you won't need to determine the sysdig/falco version your code is compatible with
> 
> @zuc sure, but does that affect this proposal?

Absolutely, that's why I'm excited about the proposal so I stop having to do it.  My comment more was in response to your comment about versioning number.  Just to say, I think you are fairly safe setting a version number that is higher than falco/sysdig but lower than 1.0.0 if you want.

**Comment by gnosek** [comment]
@terylt I believe starting at 1.0.0 makes sense (the API has been stable for years now with minor changes). We're about to release 12.0.0, so we can't just replace the product version with the API version in the prebuilt module names but I hope we'll be able to just change the module name (to `scap`).

Or we start at MAJOR_VERSION=32. 🤷

File: proposals/20210818-driver-semver.md
Comment (terylt) [reply proposals/20210818-driver-semver.md]: Yes, it doesn't have to be exactly that `falco-driver-loader` file, but something that does something similar.  Either builds the ebpf/kernel probe from source or downloads from a pre-built `scap` probe location - that could be configurable if a consumer wants to use its own pre-built probe repo.  

All the docker-entrypoint.sh does is mount the kernel header/src directories to /host/ in the container and then calls the aforementioned driver loader.   When porting the sysflow collector over from the sysdig libs to the falco libs, I still needed to link to the falco project proper for three main reasons:
1.) Copy over the falco-driver-loader file to automatically build or download the prebuilt probes.
2.) Copy over the docker-entrypoint.sh to setup the environment so the probe could be built properly and the driver loader called,
3.) the falco-libs patch file which renamed SYSDIG_HOST_ROOT to HOST_ROOT and .sysdig to .falco

Consumers could wrap this core functionality with whatever added features they wanted.  Just a thought.

**Comment by terylt** [comment]
> @terylt I believe starting at 1.0.0 makes sense (the API has been stable for years now with minor changes). We're about to release 12.0.0, so we can't just replace the product version with the API version in the prebuilt module names but I hope we'll be able to just change the module name (to `scap`).
> 
> Or we start at MAJOR_VERSION=32. 🤷

@gnosek  The probe has been around for a number of years already so starting at 1.0.0 makes sense.

File: proposals/20210818-driver-semver.md
Comment (gnosek) [reply proposals/20210818-driver-semver.md]: Will have to think about it, but I think it could live as a separate PR/proposal, independent of this one.

I like the idea, I'd just like to avoid scope creep

**Comment by gnosek** [comment]
@leogr what are the next steps now?

File: proposals/20210818-driver-semver.md
Comment (leogr) [reply proposals/20210818-driver-semver.md]: > 3.) the falco-libs patch file which renamed SYSDIG_HOST_ROOT to HOST_ROOT and .sysdig to .falco

This point and the defaulted canonical name should be addressed by https://github.com/falcosecurity/libs/pull/41 (N.B. the PR still use `falco` as default, but I will be happy to change it to `scap` once this proposal has been approved)

Regarding other points, I agree with @gnosek : another independent proposal would be better.

Just my 2 cents

**Comment by leogr** [comment]
> @leogr what are the next steps now?

The proposal SGTM already :+1: 
It's nice and really needed, IMHO. Also, thank you for having incorporated my suggestions.

That being said, before giving my review approval, I want to wait for another couple of days to be sure a sufficient discussion has taken place and give other @falcosecurity/maintainers the time to take a look. I'd advise presenting this proposal in the next community call too.

**Review by ldegio — APPROVED** [review]

**Review by leogr — APPROVED** [review]
LGTM :+1:

**Pull request merged**
